extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergCatalogAdapter.java

```diff
@@ -860,8 +835,456 @@ private Table readTableInternal(
+                Map.of(TableProperties.DEFAULT_FILE_FORMAT, DEFAULT_GENERATED_FILE_FORMAT,
+                        TableProperties.PARQUET_COMPRESSION,
+                        StringUtils.toLowerCase(writeInstructions.compressionCodecName())));
+    }
+
+    private static class CompletedParquetWrite {
+        private final URI destination;
+        private final long numRows;
+        private final long numBytes;
+
+        private CompletedParquetWrite(final URI destination, final long numRows, final long numBytes) {
+            this.destination = destination;
+            this.numRows = numRows;
+            this.numBytes = numBytes;
+        }
+    }
+
+    @NotNull
+    private static List<CompletedParquetWrite> writeParquet(
+            @NotNull final org.apache.iceberg.Table icebergTable,
+            @NotNull final Table[] dhTables,
+            @NotNull final IcebergParquetWriteInstructions writeInstructions) {
+        // Build the parquet instructions
+        final List<CompletedParquetWrite> parquetFilesWritten = new ArrayList<>(dhTables.length);
+        final ParquetInstructions.OnWriteCompleted onWriteCompleted =
+                (destination, numRows, numBytes) -> parquetFilesWritten
+                        .add(new CompletedParquetWrite(destination, numRows, numBytes));
+        final ParquetInstructions parquetInstructions = writeInstructions.toParquetInstructions(
+                onWriteCompleted, icebergTable.schema().idToName());
+
+        // Write the data to parquet files
+        int count = 0;
+        for (final Table dhTable : dhTables) {
+            final String filename = String.format(
+                    "00000-%d-%s.parquet",
+                    count++,
+                    UUID.randomUUID());
+            final String newDataLocation = icebergTable.locationProvider().newDataLocation(filename);
+            ParquetTools.writeTable(dhTable, newDataLocation, parquetInstructions);
+        }
+        return parquetFilesWritten;
+    }
+
+    /**
+     * Commit the changes to the Iceberg table by creating snapshots.
+     */
+    private static void commit(
+            @NotNull final org.apache.iceberg.Table icebergTable,
+            @NotNull final SpecAndSchema newSpecAndSchema,
+            @NotNull final Iterable<DataFile> appendFiles,
+            final boolean overwrite,
+            final boolean schemaVerified) {
+        // Append new data files to the table
+        final Transaction icebergTransaction = icebergTable.newTransaction();
+
+        if (overwrite) {
+            // Delete all the existing data files in the table
+            final DeleteFiles deletes = icebergTransaction.newDelete();
+            try (final Stream<DataFile> dataFiles = allDataFiles(icebergTable, icebergTable.currentSnapshot())) {
+                dataFiles.forEach(deletes::deleteFile);
+            }
+            deletes.commit();
```

Comment:
I don't understand this. We should be able to, for example, compact two files into one transactionally.

Reply:
Good catch.